### ISSUE TITLE
Add force switch for resource / stack removal

### ISSUE THREAD

**Issue by pmuens** [serverless/serverless#1609]
Stack removal fails e.g. if a bucket in the stack has items in it.

We should check for exisiting items in the S3 buckets and prompt that the user can not remove the stack because of the existing items in the buckets.

The `--force` or `-f` switch should be available so that we remove the items in the bucket and the removal process succeeds.

/cc @eahefnawy @flomotlik

**Comment by rowanu** [comment]
Just ran in to this issue myself, and wanted to clarify: The other option with CFN is to set a [`"DeletionPolicy": "Retain"`](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html) on the bucket, so that the stack can be deleted successfully, but no data is lost. Obviously there's an on-going cost (however small) with storing things in S3, so I'm not sure if that's desirable.

What's your preference in this case? I feel like `--force` should delete **everything** as that's what it normally does (e.g. `rm -rf`).

**Comment by pouledodue** [comment]
@rowanu 'retain' option would be great, why not offer both option

**Comment by pmuens** [comment]
@rowanu that's also an interesting setting.

The `--force` switch should enable the user to completely remove all resources related to the service (we want to reduce time you'll otherwise spend in the AWS console).

Do you guys have ideas for other services which might cause problems if one wants to remove the whole service? Or is S3 one of the only services which might fail when not properly prepared for stack removal?

**Comment by rowanu** [comment]
The policy is used to retain state after resources have been cleaned-up and so many of it's use-cases are aimed at stateful servers (e.g. volumes and snapshots), which is obviously not an issue with Serverless. The only other example that comes to mind is DynamoDb - if DDB tables are ever provisioned as part of a deployment you might not want to automatically delete them.

**Comment by pouledodue** [comment]
retain DDB data is a must IMHO.

**Comment by andymac4182** [comment]
If you have changed a IamRole by attaching a managed policy it will cause the Delete Stack to fail.

**Comment by pmuens** [comment]
Thanks @andymac4182 that's good to know.

So this feature might need a little more in-depth investigation before starting the implementation phase.

Maybe we can start with a simple warning which gets printed on the console before the StackRemoval process starts and also document the whole process / pitfalls.

**Comment by dougmoscrop** [comment]
not sure where to post this, but I noticed that apex supports a command called 'infra'. So the idea is you don't provision your database with the serverless.yml resources, you set up something before (i.e. terraform) that does it. I don't know how you feel about this but I think it makes sense in a 'layering' sense. serverless remove should tear down and remove the things laid out with serverless, and a command like this (a separate plugin - maybe even one for cf, one for tf, etc.) could be responsible for stuff like dbs/persistence.

**Comment by rowanu** [comment]
Not sure that fits with the Serverless approach so far. There's currently no distinction between resources created "under the hood" for you (e.g. Lambda functions) and resource you create yourself in the `resources` section (e.g. DDB tables); When you remove a service, you remove all the resources created by it (which I think is a good default position to have, due to cost, etc).

I think if you want "long lived" resources that are part of *multiple* Serverless services (like data persistence) then they shouldn't be managed by Serverless _at all_ (at least not in its current form).

I know there's been discussion about the idea of "related" services, but that's not in the scope of this issue; This issue is just regarding what we do _by default_ to services that might have data persistence.

**Comment by andymac4182** [comment]
This is where cross stack references would probably be good if we could use them to populate variables that the lambda code could access. You could have a long lived stack and api one. Would probably also solve how to handle multiple services behind one gateway or multiple stages behind a single gateway.

**Comment by flomotlik** [comment]
@andymac4182 yup that was my idea as well. And btw as we deploy straight to CF this is already possible. Anything you add into resrouces->(Resources|Outputs) will be merged into the CF template, so all those new options are immediately open to you.

**Comment by flomotlik** [comment]
After thinking about it more I don't think we should implement a force switch. Users have the ability to add different policies through custom resources already and removing resources is just a bit too dangerous imho.
